vignettes/collaboration.Rmd

```diff
@@ -0,0 +1,261 @@
+
+In most of the situaions above, the users are never holding the canonical source of the packets - so it does not matter if they don't have all if their dependencies. They do hold all the metadata for all packets (again, recursively) and could download the dependencies later if they wanted to inspect them. Typically though, this would just be extra space used up on disk.
+
+When you push a packet to a server, it always pushes the complete tree.
+
+# Moving as little data as possible
+
+When pulling packets, first the user pulls metadata from the server. When `orderly2::orderly_location_metadata_pull` is run, all metadata from the server is updated on the client (mrc-4444 wil make this lazier). We now know everything the server holds, and the hashes of all the files contained in each packet.
+
+It is very likely that we already know of some of the files within a packet being requested, so the client first looks to see if these exist within its archive, verifying that they have not been changed.  It then requests from the server all the files that it cannot resolve locally. If doing a recursive pull (with `core.require_complete_tree = TRUE`) then we look at the union over all files in the set of packets that are currently missing from our archive. This is frequently much less than the full set of files within the packets.
+
+The algorithm is similar for a push:
+
+* look at the full dependency chain of our target packet and find all file ids and ask the server which if these it does not yet know about
+* in the unknown subset we get the union of file hashes and ask the server which subset of files it does not yet know about
+* we then push up files - this is done one at a time at present but could be done simultaneously with async requests in future
+* once all files are on the sever we upload the missing metadata, in chronological order and the server verifies that for each packet it has metadata for all dependencies and all referenced files
+
+This multi-step process means that we avoid copying data that is already known about, as well as avoiding moving the same data multiple times. It also means that if the process of pushing (or pulling) is interrupted it can be safely resumed as the state is always consistent.
+
+# Interaction between source tree and `.outpack`
+
+It is important that the `.outpack` directory is *not* shared via git; we warn about this now, and you can use `orderly2::ordery_gitignore_update()` to automatically create a suitable `.gitignore` file that will prevent it being accidentically committed.
+
+However, 
+
+If Alice and Bob were starting on a new machine they would:
+
+* clone their source repository
+* run `orderly::orderly_init()` from within the directory
```

Comment:
This ends fairly abruptly? Was there a final point to make? If they run `orderly::orderly_init()` from within the directory it will create the `.outpack` directory?

Reply:
I've expanded this somewhat now